sdk/secp256r1/src/lib.rs

```diff
@@ -0,0 +1,662 @@
+            solana_logger::setup();
+            let message_arr = b"hello";
+            let group = EcGroup::from_curve_name(Nid::X9_62_PRIME256V1).unwrap();
+            let signing_key = EcKey::generate(&group).unwrap();
+            let instruction = new_secp256r1_instruction(message_arr, signing_key).unwrap();
+            let mint_keypair = Keypair::new();
+            let feature_set = FeatureSet::all_enabled();
+
+            let tx = Transaction::new_signed_with_payer(
+                &[instruction.clone()],
+                Some(&mint_keypair.pubkey()),
+                &[&mint_keypair],
+                Hash::default(),
+            );
+
+            assert!(tx.verify_precompiles(&feature_set).is_ok());
+
+            let mut instruction = instruction;
+            let index = loop {
+                let index = thread_rng().gen_range(0, instruction.data.len());
+                if index != 1 {
+                    break index;
+                }
+            };
+
+            instruction.data[index] = instruction.data[index].wrapping_add(12);
+            let tx = Transaction::new_signed_with_payer(
+                &[instruction],
+                Some(&mint_keypair.pubkey()),
+                &[&mint_keypair],
+                Hash::default(),
+            );
+            assert!(tx.verify_precompiles(&feature_set).is_err());
+        }
+
+        #[test]
+        fn test_secp256r1_high_s() {
+            solana_logger::setup();
+            let message_arr = b"hello";
+            let group = EcGroup::from_curve_name(Nid::X9_62_PRIME256V1).unwrap();
+            let signing_key = EcKey::generate(&group).unwrap();
+            let mut instruction = new_secp256r1_instruction(message_arr, signing_key).unwrap();
+
+            // Get the signature offset and modify the S value
+            let signature_offset = DATA_START + COMPRESSED_PUBKEY_SERIALIZED_SIZE;
+            let s_offset = signature_offset + FIELD_SIZE; // S comes after R
+
+            // Create a high S value by doing order - s
+            let order = BigNum::from_slice(&SECP256R1_ORDER).unwrap();
+            let current_s =
+                BigNum::from_slice(&instruction.data[s_offset..s_offset + FIELD_SIZE]).unwrap();
+            let mut high_s = BigNum::new().unwrap();
+            high_s.checked_sub(&order, &current_s).unwrap();
+
+            // Replace the S value in the signature with our high S
+            instruction.data[s_offset..s_offset + FIELD_SIZE].copy_from_slice(&high_s.to_vec());
+
+            let mint_keypair = Keypair::new();
+            let feature_set = FeatureSet::all_enabled();
+            let tx = Transaction::new_signed_with_payer(
+                &[instruction],
+                Some(&mint_keypair.pubkey()),
+                &[&mint_keypair],
+                Hash::default(),
+            );
+
+            // Should fail verification due to high s value
+            assert!(tx.verify_precompiles(&feature_set).is_err());
```

Comment:
Can we `unwrap_err()` here and test for the precise error type?

Reply:
Really sorry for the late reply on this 🙏 . I think it is fine to just have `InvalidSignature` and we can just check for this generic case. If we define a new variant `PrecompileError::InvalidSignatureRange`, then now we technically have a more specific variant of `InvalidSignature` and keeping both variants is a duplication.